internal/reader/onchain_prices_reader.go

```diff
@@ -84,5 +89,57 @@ func (pr *OnchainTokenPricesReader) GetTokenPricesUSD(
 	return prices, nil
 }
 
+func (pr *OnchainTokenPricesReader) getRawTokenPrice(ctx context.Context, token types.Account) (*big.Int, error) {
+	var latestRoundData *LatestRoundData
+	if err :=
+		pr.ContractReader.GetLatestValue(
+			ctx,
+			consts.ContractNamePriceAggregator,
+			consts.MethodNameGetLatestRoundData,
+			primitives.Unconfirmed,
+			nil,
+			&latestRoundData,
+			//boundContract,
+		); err != nil {
+		return nil, fmt.Errorf("latestRoundData call failed for token %s: %w", token, err)
+	}
+
+	if latestRoundData == nil {
+		return nil, fmt.Errorf("latestRoundData is nil for token %s", token)
+	}
+
+	return latestRoundData.Answer, nil
+}
+
+func (pr *OnchainTokenPricesReader) getTokenDecimals(ctx context.Context, token types.Account) (*uint8, error) {
+	var decimals *uint8
+	if err :=
+		pr.ContractReader.GetLatestValue(
+			ctx,
+			consts.ContractNamePriceAggregator,
+			consts.MethodNameGetDecimals,
+			primitives.Unconfirmed,
+			nil,
+			&decimals,
+			//boundContract,
+		); err != nil {
+		return nil, fmt.Errorf("decimals call failed for token %s: %w", token, err)
+	}
+
+	return decimals, nil
+}
+
+// Input price is USD per full token, with 18 decimal precision
+// Result price is USD per 1e18 of smallest token denomination, with 18 decimal precision
+// Examples:
+//
+//	1 USDC = 1.00 USD per full token, each full token is 1e6 units -> 1 * 1e18 * 1e18 / 1e6 = 1e30
+//	1 ETH = 2,000 USD per full token, each full token is 1e18 units -> 2000 * 1e18 * 1e18 / 1e18 = 2_000e18
+//	1 LINK = 5.00 USD per full token, each full token is 1e18 units -> 5 * 1e18 * 1e18 / 1e18 = 5e18
+func calculateUsdPer1e18TokenAmount(price *big.Int, decimals uint8) *big.Int {
```

Comment:
Hmm the token price isn't always to 18 decimals though, that depends on the feed.

Reply:
You need to normalize the price to 18 decimals before you call this function, see https://github.com/smartcontractkit/ccip/blob/5e1b6d179ca8428b52a025597381c32edecf37a7/core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go#L175-L184

Reply:
That is what the function is doing. It also shows in the tests, it's normalizing to 18 decimals depending on the decimals sent as a parameter.